x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/use_prebuilt_rules_upgrade_state.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useMemo, useState } from 'react';
+import {
+  ThreeWayDiffConflict,
+  type DiffableAllFields,
+  type DiffableRule,
+  type RuleObjectId,
+  type RuleSignatureId,
+  type RuleUpgradeInfoForReview,
+} from '../../../../../../common/api/detection_engine';
+import { convertRuleToDiffable } from '../../../../../../common/detection_engine/prebuilt_rules/diff/convert_rule_to_diffable';
+
+export interface RuleUpgradeState extends RuleUpgradeInfoForReview {
+  /**
+   * Rule containing desired values users expect to see in the upgraded rule.
+   */
+  finalRule: DiffableRule;
+  /**
+   * Indicates whether there are conflicts blocking rule upgrading.
+   */
+  hasUnresolvedConflicts: boolean;
+}
+export type RulesUpgradeState = Record<RuleSignatureId, RuleUpgradeState>;
+export type SetFieldResolvedValueFn<
+  FieldName extends keyof DiffableAllFields = keyof DiffableAllFields
+> = (params: {
+  ruleId: RuleObjectId;
+  fieldName: FieldName;
+  resolvedValue: DiffableAllFields[FieldName];
+}) => void;
+
+type RuleResolvedConflicts = Partial<DiffableAllFields>;
+type RulesResolvedConflicts = Record<string, RuleResolvedConflicts>;
+
+interface UseRulesUpgradeStateResult {
+  rulesUpgradeState: RulesUpgradeState;
+  setFieldResolvedValue: SetFieldResolvedValueFn;
+}
+
+export function usePrebuiltRulesUpgradeState(
+  ruleUpgradeInfos: RuleUpgradeInfoForReview[]
+): UseRulesUpgradeStateResult {
+  const [rulesResolvedConflicts, setRulesResolvedConflicts] = useState<RulesResolvedConflicts>({});
+  const setFieldResolvedValue = useCallback((...[params]: Parameters<SetFieldResolvedValueFn>) => {
+    setRulesResolvedConflicts((prevRulesResolvedConflicts) => ({
+      ...prevRulesResolvedConflicts,
+      [params.ruleId]: {
+        ...(prevRulesResolvedConflicts[params.ruleId] ?? {}),
+        [params.fieldName]: params.resolvedValue,
+      },
+    }));
+  }, []);
+  const rulesUpgradeState = useMemo(() => {
+    const state: RulesUpgradeState = {};
+
+    for (const ruleUpgradeInfo of ruleUpgradeInfos) {
+      state[ruleUpgradeInfo.rule_id] = {
+        ...ruleUpgradeInfo,
+        finalRule: calcFinalDiffableRule(
+          ruleUpgradeInfo,
+          rulesResolvedConflicts[ruleUpgradeInfo.id] ?? {}
```

Comment:
If understand it correctly, we should use `ruleUpgradeInfo.rule_id` here and when calling `calcUnresolvedConflicts` below. Otherwise it won't find an item in `rulesResolvedConflicts`.